### PR TITLE
Fix TypeScript usage with styled-components

### LINKED ISSUE #23: TypeScript CSSObject as String error
# tl;dr
If using `twin.macro` inside a Styled Components template literal you get typescript errors.

# Basic Example

```typescript
const Submit = styled(SubmitButton)<submitButtonProps>`
  ${tw`w-full`}
`
```

then you get errors like these

```
Argument of type 'CSSObject' is not assignable to parameter of type 'Interpolation<any>'.
  Type 'CSSObject' is not assignable to type 'string'.
```

# Complex Example
A more complex example:

```typescript
const Container = styled.div`
  ${tw`flex flex-col max-w-sm w-full`}
  & > form {
    ${tw`w-full`}
  }
`
```

gives you the error

```
No overload matches this call.
  Overload 1 of 3, '(first: TemplateStringsArray | CSSObject | InterpolationFunction<ThemedStyledProps<Pick<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, "style" | ... 254 more ... | "css"> & { ...; }, any>>, ...rest: Interpolation<...>[]): StyledComponent<...>', gave the following error.
    Argument of type 'CSSObject' is not assignable to parameter of type 'Interpolation<ThemedStyledProps<Pick<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, "style" | "title" | "slot" | ... 252 more ... | "css"> & { ...; }, any>>'.
      Type 'CSSObject' is not assignable to type 'string'.
  Overload 2 of 3, '(first: TemplateStringsArray | CSSObject | InterpolationFunction<ThemedStyledProps<Pick<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, "style" | ... 254 more ... | "css"> & { ...; } & object, any>>, ...rest: Interpolation<...>[]): StyledComponent<...>', gave the following error.
    Argument of type 'CSSObject' is not assignable to parameter of type 'Interpolation<ThemedStyledProps<Pick<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, "style" | "title" | "slot" | ... 252 more ... | "css"> & { ...; } & object, any>>'.
      Type 'CSSObject' is not assignable to type 'string'.
```

### PR DESCRIPTION
**Pull request by kingdaro**
Closes #23

### CODE REVIEW

**Comment by rluiten** [comment]
I noticed some problems with alpha8 typescript myself.
Just had a look through this.

Have you considered the follow test case for the types?

```typescript
const DivBgWhite = tw.div`bg-white`
const TestTwDiv = () => <div><DivBgWhite>Words here</DivBgWhite></div>
```

I grabbed the source of this PR and added the above test to the test.tsx and it fails with error.
```
JSX element type 'DivBgWhite' does not have any construct or call signatures.ts(2604)
````

I had a look at how emotion sets the types.

@emotion/styled/types/index.d.ts  (from version 10.0.27)
It uses a type StyledTags<>

https://github.com/emotion-js/emotion/blob/master/packages/styled/types/index.d.ts#L24
image

**Comment by kingdaro** [comment]
Ah, missed that 😅 I only use the regular `tw` form myself

**Comment by rbutera** [comment]
@ben-rogerson need any help with reviewing/merging this PR?

**Comment by rbutera** [comment]
@kingdaro will this PR fix #33?

**Comment by ben-rogerson** [comment]
> @ben-rogerson need any help with reviewing/merging this PR?

that would be really appreciated 👍

**Comment by kingdaro** [comment]
> @kingdaro will this PR fix #33?

@rbutera It should. [There's a test for the tw prop, but I made it slightly more comprehensive to be sure](https://github.com/ben-rogerson/twin.macro/pull/24/files#diff-6d48f4c6e3078513e68c2c416ea404b8R39-R43)

**Comment by rbutera** [comment]
@ben-rogerson if you could make me a contributor on GitHub and assign this PR to me then I'll do a review ASAP!

I suggest also making @kingdaro a contributor if he wants?

**Comment by kingdaro** [comment]
> I suggest also making @kingdaro a contributor if he wants?

Sure, wouldn't mind that 🙂

**Comment by ben-rogerson** [comment]
Daro is already a Twin contributor because of his typescript pr's 👍 
Do you mean collaborator? That would grant the ability to set you as a reviewer but comes with a [whole lot of other privileges](https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/permission-levels-for-a-user-account-repository#collaborator-access-on-a-repository-owned-by-a-user-account).
If either of you are keen, [send me a twitter pm](https://twitter.com/benrogerson).

**Comment by kingdaro** [comment]
I could only ever work with TS-related PRs since I can't work with macros, or any of that language transformation stuff 😅 I'll pass on collaborator

**Comment by rbutera** [comment]
@kingdaro 

This PR looks really good. I'm excited to finally get rid of those red squiggles for styled-components.

However please see my comment. There is a problem with `types/tests/emotion.tsx`

File: types/tests/emotion.tsx [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import tw from '..'
+
+export const Container = styled.div`
+  ${tw`flex flex-col w-full max-w-sm`}
+  & > form {
+    ${tw`w-full`}
+  }
+`
+
+export const cssProp = <div css={tw`bg-red-100`} />
```
Comment (rbutera) [inline types/tests/emotion.tsx:13]: typescript error on this line

```
Type '{ css: TwStyle; }' is not assignable to type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.
  Property 'css' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.
```

File: types/tests/styled-components.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+export const Container = styled.div`
+  ${tw`flex flex-col w-full max-w-sm`}
+  & > form {
+    ${tw`w-full`}
+  }
+`
+
+// I want to test the styled-components css prop in isolation from the emotion css prop,
+// but I dunno how to do that
+// export const cssProp = <div css={tw`bg-red-100`} />
```
Comment (rbutera) [inline types/tests/styled-components.tsx:13]: wouldn't this prop be taken from the import of `styled-components`? 

I assume that the emotion `css` prop is only injected on files processed by emotion's `jsx`

**Comment by rbutera** [comment]
just noticed it may be due to missing emotion types? 

image

File: types/tests/emotion.tsx (hunk 1 above)
Comment (kingdaro) [reply types/tests/emotion.tsx:13]: I'm not getting a type error in my editor here, and `npm run test:types` doesn't show anything. Make sure you're using the workspace version of TS (it's using the 3.9 beta for `// @ts-expect-error`)

File: types/tests/styled-components.tsx (hunk 2 above)
Comment (kingdaro) [reply types/tests/styled-components.tsx:13]: Emotion's `css` prop is a global type declaration on the `JSX` namespace. Uncomment that line, then ctrl+click on `css`, and you'll see it leads to emotion's types, not SC's, and I'm not sure how to isolate the global typings

**Comment by kingdaro** [comment]
🤔 I have them listed as dev deps, so I'm not sure what's up there

File: types/tests/styled-components.tsx (hunk 2 above)
Comment (kingdaro) [reply types/tests/styled-components.tsx:16]: Strike that, it looks like I just needed to put each test in their own folder with their own tsconfig. Kind of noisy, but eh, it works

**Comment by nextriot** [comment]
Any ETA on this getting merged? :)

**Comment by ben-rogerson** [comment]
Ok I'll hurry this one along. Expect a release within 24 hours 👍

**Comment by rbutera** [comment]
@ben-rogerson happy Easter. I suggest you merge 34 first so we can benefit from the code formatting etc

Having reviewed this I think it’s ready to be merged. I fixed my missing types error by rerunning yarn

**Comment by ben-rogerson** [comment]
Sorry I should have asked, great to hear it's good to go

**Comment by ben-rogerson** [comment]
Big thanks to @kingdaro for beefing up the types and @rbutera for the review 🙏

**Pull request merged**